### PR TITLE
Add and train convolutional neural network model 

### PR DESCRIPTION
**Pull request by piotrwinkler**
This script loads data from previously prepared data server and then creates a convolutional neural network model based on keras. The model is trained on provided data and the results are saved in project folder as ".hdf5" files. In the #Normalization section there is also a part that limits data to small values which are better processed by neural networks. At the end there is a loop which loads previously saved model weights and tests them on testing dataset.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -1 +1,78 @@
...
+import matplotlib.pyplot as plt
+from keras.models import Sequential
+from keras.models import load_model
+from keras.layers import Dense, Dropout, Activation, Flatten
+from keras.layers import Convolution2D, MaxPooling2D
+from keras.utils import np_utils
+from keras import optimizers
+import sys
+from keras.callbacks import ModelCheckpoint
+sys.path.append('\PycharmProjects\smile-warrior')
```
Comment (pawelkopec) [inline train.py]: This should be set with PYTHONPATH, not within script. Additionally not everybody has PycharmProjects directory.

File: train.py [hunk 2]
```diff
@@ -1 +1,78 @@
...
+from data_server import *
```
Comment (pawelkopec) [inline train.py]: Good practice is to avoid * and specify what you import.

File: train.py [hunk 3]
```diff
@@ -1 +1,78 @@
...
+
+
+#Downloading and reshaping data
+X_train, Y_train, X_test, Y_test, X_validate, Y_validate = Load_dataset('smile_warrior_dataset.csv')
```
Comment (pawelkopec) [inline train.py]: Please put the path to command line arguments.

File: train.py [hunk 4]
```diff
@@ -1 +1,78 @@
...
+X_train = X_train.reshape(X_train.shape[0], 48, 48, 1)
```
Comment (pawelkopec) [inline train.py]: You do not need to hardcode 48. You can just run sqrt of X_train dimension.

File: train.py [hunk 5]
```diff
@@ -1 +1,78 @@
...
+X_test = X_test.reshape(X_test.shape[0], 48, 48, 1)
+X_validate = X_validate.reshape(X_validate.shape[0], 48, 48, 1)
+
+
+#Normalization
+mean_image = np.mean(X_train)
```
Comment (pawelkopec) [inline train.py]: If you extract mean and std, you need to save it to reuse it for inference.
```
normalized = (image - 256/2)/256 # This will work well too without additional params
```

File: train.py (hunk 3 above)
Comment (pawelkopec) [inline train.py]: General comment - names of variables from lower case in python.

File: train.py [hunk 6]
```diff
@@ -1 +1,78 @@
...
+X_validate = (X_validate - mean_image) / std_image
+
+
+#Preparing a proper format of output data
+Y_train = np_utils.to_categorical(Y_train, 2)
+Y_test = np_utils.to_categorical(Y_test, 2)
+Y_validate = np_utils.to_categorical(Y_validate, 2)
+
+
+#Network parameters
```
Comment (pawelkopec) [inline train.py]: Also to command line arguments. You can add those values as default, but it should be easy to change.

File: train.py [hunk 7]
```diff
@@ -1 +1,78 @@
...
+from keras.layers import Dense, Dropout, Activation, Flatten
+from keras.layers import Convolution2D, MaxPooling2D
+from keras.utils import np_utils
+from keras import optimizers
+import sys
+from keras.callbacks import ModelCheckpoint
+sys.path.append('\PycharmProjects\smile-warrior')
+from data_server import *
+
+
```
Comment (pawelkopec) [inline train.py:11]: General comment - please put all this stuff into some functions.

File: train.py [hunk 8]
```diff
@@ -1 +1,78 @@
...
+model.add(Convolution2D(32, (3, 3), activation = 'relu'))
+model.add(MaxPooling2D(pool_size = (2,2)))
+model.add(Dropout(0.25))
+model.add(Flatten())
+model.add(Dense(128, activation ='relu'))
+model.add(Dropout(0.5))
+model.add(Dense(2, activation ='softmax'))
+opt = optimizers.Adam(lr = Learning_rate)
+model.compile(loss = 'categorical_crossentropy', optimizer = opt, metrics = ['accuracy'])
+checkpointer = ModelCheckpoint(filepath = 'D:\Python3.6.6\Jupyter\smile_warrior\weights.{epoch:02d}.hdf5',
```
Comment (pawelkopec) [inline train.py]: Do not hardcode the path. This will not work on my machine.

File: train.py [hunk 9]
```diff
@@ -1 +1,78 @@
...
+#Network parameters
+Batch_size = 32
+Epochs = 5
+Saving_period = 1
+Learning_rate = 0.001
+
+
+#Preparing neural network model
+model = Sequential()
+model.add(Convolution2D(32, (3, 3), activation = 'relu', input_shape = (48,48,1)))
```
Comment (pawelkopec) [inline train.py]: Again - 48 is hardcoded. Please infer it from the data and save to some variable. If we will change the dataset or image size you will have to change it in multiple places.

File: train.py [hunk 10]
```diff
@@ -1 +1,78 @@
...
+model.add(Dense(128, activation ='relu'))
+model.add(Dropout(0.5))
+model.add(Dense(2, activation ='softmax'))
+opt = optimizers.Adam(lr = Learning_rate)
+model.compile(loss = 'categorical_crossentropy', optimizer = opt, metrics = ['accuracy'])
+checkpointer = ModelCheckpoint(filepath = 'D:\Python3.6.6\Jupyter\smile_warrior\weights.{epoch:02d}.hdf5',
+                               verbose = 1, save_best_only = False, period = Saving_period)
+
+#Saving model
+model.save('model.hdf5')
```
Comment (pawelkopec) [inline train.py]: Why save randomly initialized model?

File: train.py [hunk 11]
```diff
@@ -1 +1,78 @@
...
+          validation_data = (X_validate, Y_validate), callbacks = [checkpointer])
+
+
+#Downloading model
+#model = load_model('model.hdf5')
+
+
+for i in range((int)(Epochs/Saving_period)):
+    #Downloading weights
+    model.load_weights("D:\Python3.6.6\Jupyter\smile_warrior\weights.0{}.hdf5".format(i+1))
```
Comment (pawelkopec) [inline train.py]: Hardcoded path again.

Comment (pawelkopec) [inline train.py]: This looks wrong. You need to wait for the training to end to have the evaluation. To have validation in real time you can use callbacks of fit() method:
https://keras.rstudio.com/reference/fit.html

File: train.py [hunk 12]
```diff
@@ -1 +1,218 @@
...
+    """
+    # Downloading and reshaping data
+    x_train, y_train, x_test, y_test, x_validate, y_validate = load_dataset(args.dataset)
+    shape = int(math.sqrt(x_train.shape[1]))
+    x_train = x_train.reshape(x_train.shape[0], shape, shape, 1)
+    x_test = x_test.reshape(x_test.shape[0], shape, shape, 1)
+    x_validate = x_validate.reshape(x_validate.shape[0], shape, shape, 1)
+
+    # Normalization
+    x_train = (x_train - 256/2) / 256
```
Comment (pawelkopec) [inline train.py]: Please put normalisation into a function

File: train.py [hunk 13]
```diff
@@ -1 +1,218 @@
...
+
+    # Preparing a proper format of output data
+    y_train = np_utils.to_categorical(y_train, 2)
+    y_test = np_utils.to_categorical(y_test, 2)
+    y_validate = np_utils.to_categorical(y_validate, 2)
+
+    return x_train, y_train, x_test, y_test, x_validate, y_validate, shape
+
+
+def network_preparation(shape, learning_rate, saving_period, args):
```
Comment (pawelkopec) [inline train.py]: Rather use verbs eg build_network(), prepare_network()

File: train.py [hunk 14]
```diff
@@ -1 +1,218 @@
...
+    """
+    This module prepares convolutional neural network model for future training.
+
+    :param shape:
```
Comment (pawelkopec) [inline train.py:63]: Shape is a tuple (width, height). Rather call this parameter side_length or something similar.

File: train.py [hunk 15]
```diff
@@ -1 +1,218 @@
...
+        Height/width of processed images.
+
+    :param learning_rate:
+        Length of single step in gradient propagation.
+
+    :param saving_period:
+        How many epochs of training must pass before saving model.
+
+    :param args:
```
Comment (pawelkopec) [inline train.py]: Bad practice. Rather pass each param separately.

File: train.py [hunk 16]
```diff
@@ -1 +1,218 @@
...
+
+    checkpointer = ModelCheckpoint(filepath=args.models_dir,
+                                   verbose=1, save_best_only=False, save_weights_only=False, period=saving_period)
+
+    tensorboard = TensorBoard(log_dir=args.logs_dir.format(time()))
+
+    return model, checkpointer, tensorboard
+
+
+def training(batch_size, epochs, model, checkpointer, tensorboard, x_train, y_train, x_validate, y_validate):
```
Comment (pawelkopec) [inline train.py]: Again - verb better, eg train()

File: train.py [hunk 17]
```diff
@@ -1 +1,218 @@
...
+from keras.utils import np_utils
+from keras import optimizers
+from keras.callbacks import ModelCheckpoint, TensorBoard
+
+
+def data_preparation(args):
+    """
+    This module prepares data for training and validation process.
+
+    :param args:
```
Comment (pawelkopec) [inline train.py]: Please pass each arg separately.

File: train.py [hunk 18]
```diff
@@ -1 +1,218 @@
...
+    parser.add_argument('--epochs', '-e', metavar="number of training epochs",
+                        help='number of training epochs', type=int, default=10)
+
+    parser.add_argument('--saving_period', '-s', metavar="how many epochs must pass before saving",
+                        help='how many epochs must pass before saving', type=int, default=1)
+
+    parser.add_argument('--learning_rate', '-lr', metavar="specifies learning rate",
+                        help='specifies learning rate', type=float, default=0.0001)
+
+    parser.add_argument('--models_dir', '-m', metavar="specifies saving directory for models",
```
Comment (pawelkopec) [inline train.py]: Rather use - than _

Comment (piotrwinkler) [reply train.py]: Are you sure? I get errors when I try to do this that way.

File: train.py [hunk 19]
```diff
@@ -1 +1,241 @@
...
+    :param x_test:
+        Input testing data.
+
+    :param x_validate:
+        Input validation data.
+
+    :return:
+        Normalized input images.
+    """
+    x_train = (x_train - 256 / 2) / 256
```
Comment (pawelkopec) [inline train.py]: This does not remove redundancy. Rather make this function look sth like:
```
def normalize(data):
    return (data - 256 / 2) / 256

File: train.py [hunk 20]
```diff
@@ -1 +1,241 @@
...
+
+    checkpointer = ModelCheckpoint(filepath=models_dir,
+                                   verbose=1, save_best_only=False, save_weights_only=False, period=saving_period)
+
+    tensorboard = TensorBoard(log_dir=logs_dir.format(time()))
+
+    return model, checkpointer, tensorboard
+
+
+def train(batch_size, epochs, model, checkpointer, tensorboard, x_train, y_train, x_validate, y_validate):
```
Comment (pawelkopec) [inline train.py]: Why do you need this function if you just pass everything to a ready method ```model.fit()```?

File: train.py [hunk 21]
```diff
@@ -1 +1,241 @@
...
+
+    :param model
+        Tested  neural network model.
+    """
+    for i in range(int(epochs/saving_period)):
+        # Downloading weights
+        try:
+            model = load_model(load_dir.format(i+1))
+        except:
+            print("Unable to load model")
```
Comment (pawelkopec) [inline train.py]: This looks very bad. Firstly you hide an actual error and replace it with mysterious message "Unable to load model". Secondly you proceed with the model even though you caught an error.

File: train.py [hunk 22]
```diff
@@ -1 +1,241 @@
...
+def parse_args():
+    """
+    This module parses command line arguments.
+
+    :return:
+        Parser containing command line data.
+    """
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument('--dataset', '-d', metavar="csv file containing dataset",
```
Comment (pawelkopec) [inline train.py]: Why do you need all those metavars if you just repeat help message?

File: train.py [hunk 23]
```diff
@@ -1 +1,241 @@
...
+                        help='specifies path for loading models', default="model.0{}.hdf5")
+
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+
+    # Network parameters
+    batch_size = args.batch_size
```
Comment (pawelkopec) [inline train.py:141]: I would not unpack these, but I leave it to your choice.

Comment (piotrwinkler) [reply train.py:141]: I think I'll leave it as it is.
